src/gpuarray/buffer.h

```diff
@@ -757,6 +763,48 @@ typedef enum _ga_usefl {
   GA_USE_OPENCL =   0x4000,
 } ga_usefl;
 
+/************************************************************************************
+*                         multi-gpu communicator interface                         *
+************************************************************************************/
+
+/**
+ * \brief Creates a new communicator (multi process version)
+ * \param clique_id [const char*] TODO
+ * \param ndev [int] TODO
+ * \param rank [int] TODO
+ * \param res [int*] TODO
+ * \return gpucomm* TODO
+ */
+GPUARRAY_PUBLIC gpucomm* gpucomm_init(const char* clique_id, int ndev, int rank,
+                                      int* res);
+
+/**
+ * \brief Frees resources associated with communicator object.
+ * \param comm [gpucomm*] TODO
+ * \return void TODO
+ */
+GPUARRAY_PUBLIC void gpucomm_destroy(gpucomm* comm);
+
+/**
+ * \brief Returns nice error message.
+ * \param comm [gpucomm*] TODO
+ * \param err [int] TODO
+ * \return const char* TODO
+ */
+GPUARRAY_PUBLIC const char* gpucomm_error(gpucomm* comm, int err);
+
+// TODO cuda extension:
+// 1. ncclGetUniqueId
+// 2. ncclCommCount (below mby ops in gpucomm struct?!)
+// 3. ncclCommUserRank
+// 4. ncclCommCuDevice
```

Comment:
I don't believe that a call to `gpucomm_gen_clique_id` is an obligation for things to work out, because the user can provide a unique clique_id of his own creation. But, yes, most of the times it would be useful to use this helper function. To sum up we have the follow restrictions:

1. We need to provide the byte-array size with the unique byte-array id: Yes, we could provide a dummy struct of a given size along with a macro which defines the "internal" byte-array size. I prefer this approach in general, but because of the case of a possible opencl collectives implementation, it's hard to know what this constant byte-array length will be. The second option seems better in the cost of having a larger interface. We could also provide just the `const char*` along with macros for each libgpuarray predefined backend. It is user's responsibility to know which implementation is actually using anyways. I will first go for the dummy struct which will at first be just a combatible type with `ncclUniqueId`.

2. The necessity of this function: Ok I will place this function in `gpucomm`, but as it is not a crucial part of such an interface, I recommend to provide ourselves an implementation of this functionality to be used in cases where one is not provided natively by the backend (opencl may choose not to provide such a helper function, but leave it up to the user).